oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java

```diff
@@ -605,8 +814,161 @@ private void collectDeletedDocuments(GCPhases phases,
+                        .map(DocumentNodeState::getAllBundledProperties)
+                        .map(Map::keySet)
+                        .map(p -> p.stream().map(Utils::escapePropertyName).collect(toSet()))
+                        .orElse(emptySet());
+
+                final int deletedPropsGCCount = properties.stream()
+                        .filter(p -> !retainPropSet.contains(p))
+                        .mapToInt(x -> {
+                            updateOp.remove(x);
+                            return 1;})
+                        .sum();
+
+                deletedPropsCountMap.put(doc.getId(), deletedPropsGCCount);
+
+                if (log.isDebugEnabled()) {
+                    log.debug("Collected {} deleted properties for document {}", deletedPropsGCCount, doc.getId());
+                }
+                phases.stop(GCPhase.DETAILED_GC_COLLECT_PROPS);
+            }
+        }
+
+        private void collectOldRevisions(NodeDocument doc, GCPhases phases, UpdateOp updateOp) {
+
+            if (phases.start(GCPhase.DETAILED_GC_COLLECT_OLD_REVS)){
+                // TODO add old rev collection logic
+                phases.stop(GCPhase.DETAILED_GC_COLLECT_OLD_REVS);
+            }
+
+        }
+
+        int getGarbageCount() {
+            return totalGarbageDocsCount;
+        }
+
+        @Override
+        public void close() {
+            totalGarbageDocsCount = 0;
+        }
+
+        public void removeGarbage(final VersionGCStats stats) {
+
+            if (updateOpList.isEmpty()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Skipping removal of detailed garbage, cause no garbage detected");
+                }
+                return;
+            }
+
+            int updatedDocs;
+
+            monitor.info("Proceeding to update [{}] documents", updateOpList.size());
+
+            if (log.isDebugEnabled()) {
+                String collect = updateOpList.stream().map(UpdateOp::getId).collect(joining(","));
+                log.debug("Performing batch update of documents with following id's [{}]", collect);
+            }
+
+            if (cancel.get()) {
+                log.info("Aborting the removal of detailed garbage since RGC had been cancelled");
+                return;
+            }
+
+            timer.reset().start();
+            try {
+                List<NodeDocument> oldDocs = ds.findAndUpdate(NODES, updateOpList);
+                int deletedProps = oldDocs.stream().filter(Objects::nonNull).mapToInt(d -> deletedPropsCountMap.getOrDefault(d.getId(), 0)).sum();
+                updatedDocs = (int) oldDocs.stream().filter(Objects::nonNull).count();
+                stats.updatedDetailedGCDocsCount += updatedDocs;
+                stats.deletedPropsGCCount += deletedProps;
+                log.info("Updated [{}] documents, deleted [{}] properties", updatedDocs, deletedProps);
```

Comment:
Undecided about this log.info. In classic GC we do `maybeLogStats`, to reduce log frequency. This additional log.info might be noisy, wdyt?

Reply:
this would be called only after every batch of 1000 documents (and that too in case there is any garbage).

Reply:
`removeGarbage()` won't run in case there is no garbage. So it wouldn't log this message.

And in case of garbage, I think it would be better to log since we are deleting documents.